pkg/reconciler/ingress/status.go

```diff
@@ -0,0 +1,268 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"knative.dev/pkg/apis/istio/v1alpha3"
+
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	istiolisters "knative.dev/pkg/client/listers/istio/v1alpha3"
+)
+
+const (
+	probeConcurrency = 15              // defines how many probing calls can be issued simultaneously
+	stateExpiration  = 5 * time.Minute // defines how long after being last accessed a state expires
+)
+
+type probingState struct {
+	virtualService *v1alpha3.VirtualService
+	probeHost      string // the value of the HTTP 'Host' header sent when probing
+
+	pendingCount int32 // the number of pod that haven't been successfully probed yet (0 means complete success)
+	lastAccessed time.Time
+
+	context context.Context
+	cancel  func()
+}
+
+type workItem struct {
+	*probingState
+	podIP string
+}
+
+// StatusManager provides a way to check if a VirtualService is ready
+type StatusManager struct {
+	// mu guards probingStates
+	mu            sync.Mutex
+	probingStates map[string]*probingState
+
+	workQueue workqueue.RateLimitingInterface
+
+	gatewayLister    istiolisters.GatewayLister
+	podLister        corev1listers.PodLister
+	transportFactory func() http.RoundTripper
+
+	readyCallback func(*v1alpha3.VirtualService)
+}
+
+// NewStatusManager creates a new instance of StatusManager
+func NewStatusManager(
+	gatewayLister istiolisters.GatewayLister,
+	podLister corev1listers.PodLister,
+	transportFactory prober.TransportFactory,
+	readyCallback func(*v1alpha3.VirtualService)) *StatusManager {
+	return &StatusManager{
+		probingStates: make(map[string]*probingState),
+		workQueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			"ProbingQueue"),
+		gatewayLister:    gatewayLister,
+		podLister:        podLister,
+		transportFactory: transportFactory,
+		readyCallback:    readyCallback,
+	}
+}
+
+// IsReady checks if the provided VirtualService is ready, i.e. the Envoy pods serving the VirtualService
+// have all been updated. This function is designed to be used by the ClusterIngress controller, i.e. it
+// will be called in the order of reconciliation. This means that if IsReady is called on a VirtualService,
+// this VirtualService is the latest know version and therefore anything related to older versions can be ignored.
+// Also, it means that IsReady is not called concurrently.
+func (m *StatusManager) IsReady(vs *v1alpha3.VirtualService) (bool, error) {
+	key := fmt.Sprintf("%s/%s", vs.Namespace, vs.Name)
+
+	// Find the probe host
+	var probeHost string
+	for _, host := range vs.Spec.Hosts {
+		if strings.HasSuffix(host, resources.ProbeHostSuffix) {
+			if probeHost != "" {
+				return false, fmt.Errorf("only one probe host can be defined in a VirtualService, found at least 2: %s, %s", probeHost, host)
+			}
+			probeHost = host
+		}
+	}
+	if probeHost == "" {
+		return false, fmt.Errorf("only a VirtualService with a probe can be probed. Hosts: %v", vs.Spec.Hosts)
```

Comment:
returning whole set in the error probably doesn't make much sense. But log it instead.

Reply:
How so? This error will be logged by the controller anyway.
If we can avoid adding logging dependency, it would be great.

Reply:
This is a very long error with all the hosts in it. That should not be part of the error, but rather a log statement that can be correlated with the error.

Reply:
This error will be logged by the controller anyway.
What is the rational between logging and returning an error?